### PR TITLE
[Issue 242][pulsar-client-go] feature: add send timeout 

### LINKED ISSUE #242: send add timeout
#### Expected behavior

I want add timeout on producer send func, the send func need pass a context parameter，so I pass a WithTimeout context 

#### Actual behavior

I have test and find that the WithTimeout context seem not work on producer send func.The producer send func not return after timeout.

#### my code

`

	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel()

	id, err := s.config.DcSyncProducer.Send(ctx, msg)
	if err != nil{
		...
	}
`
above Send func is interface Producer Send method
`

	// Send a message
	// This call will be blocking until is successfully acknowledged by the Pulsar broker.
	// Example:
	// producer.Send(ctx, pulsar.ProducerMessage{ Payload: myPayload })
	Send(context.Context, *ProducerMessage) (MessageID, error)

	// SendAsync a message in asynchronous mode
	// The callback will report back the message being published and
	// the eventual error in publishing
	SendAsync(context.Context, *ProducerMessage, func(MessageID, *ProducerMessage, error))

`

#### System configuration
Pulsar version: 2.5.0
pulsar-client-go: lastest

### PR DESCRIPTION
**Pull request by WJL3333**
Fixes #242 

### Motivation
add send timeout

### Modifications

1. ProduceOptions
add SendTimeoutCheckInterval in ProducerOptions

2. BlockingQueue
unexported BlockingQueue.Iterator
and add methods to avoid race condition.
BlockingQueue.IterateIfNonEmpty and BlockingQueue.PollIfSatisfy

3. partitionProducer
add a ticker to check if sendRequest with context has been timeout

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

add some tiny test in producer_test.go

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: no
  - The schema:  no 
  - The default values of configurations: no
  - The wire protocol: no

### Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar/internal/blocking_queue.go [hunk 1]
```diff
@@ -34,6 +34,11 @@ type BlockingQueue interface {
 	// Poll dequeue one item, return nil if queue is empty
 	Poll() interface{}
 
+	// Poll dequeue one item if the item meets the condition,
+	// return nil,true,false if queue is empty,
+	// return nil,false,false if not satisfy.
+	PollIfSatisfy(i func(item interface{}) bool) (item interface{}, empty bool, satisfy bool)
```
Comment (merlimat) [inline pulsar/internal/blocking_queue.go:40]: This doesn't seem to be related with send timeout

**Comment by WJL3333** [comment]
hi，this pr has some bad test case，i'm working on it. any suggestions？i'm new to unit test.

**Comment by wolfstudy** [comment]
> hi, this pr has some bad test case, i'm working on it. any suggestions？i'm new to unit test.
> and i found in some case send timeout is a bit longer then expect. try to solve it too.
> @merlimat @wolfstudy @jerrypeng

Thanks @WJL3333 contribution, in unit test case, please avoid using `time.sleep()`. Because, in many cases, we have no way to know the specific situation of other people's servers and network resources.

File: pulsar/internal/blocking_queue.go [hunk 2]
```diff
@@ -173,17 +206,27 @@ func (bq *blockingQueue) Size() int {
...
-
+func (bq *blockingQueue) iterator() BlockingQueueIterator {
 	return &blockingQueueIterator{
 		bq:      bq,
 		readIdx: bq.headIdx,
 		toRead:  bq.size,
 	}
 }
 
+func (bq *blockingQueue) IterateIfNonEmpty(operation IterateFunc) {
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/internal/blocking_queue.go:217]: i think BlockingQueue is thread safe.
but if you run the following code

```go 
q := NewBlockingQueue(3)
q.Put(1)
q.Put(2)
q.Put(3)
iter := q.Iterator()
q.Poll()
q.Put(4)
```
we got 4,1,2 here. but the queue should return 2 before 4.
if some goroutine run concurrently. i think the Iterator is not woking as expected.
so i add an safe IteratorIfNonEmpty to avoid this

File: pulsar/internal/blocking_queue.go [hunk 3]
```diff
@@ -141,6 +164,16 @@ func (bq *blockingQueue) Peek() interface{} {
 	return bq.items[bq.headIdx]
 }
 
+func (bq *blockingQueue) PeekApply(consumer func(item interface{})) bool {
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/internal/blocking_queue.go:167]: i'll remove it from code.  :-）

File: pulsar/producer_partition.go [hunk 4]
```diff
@@ -343,94 +444,136 @@ func (p *partitionProducer) internalFlush(fr *flushRequest) {
...
-	wg.Wait()
-	return msgID, err
+	for {
+		select {
+		case <-wg:
+			close(wg)
+			return msgID, err
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go:480]: got it, will fix

File: pulsar/producer_partition.go [hunk 5]
```diff
@@ -343,94 +444,136 @@ func (p *partitionProducer) internalFlush(fr *flushRequest) {
...
+	wg := make(chan struct{}, 1)
 
 	var err error
 	var msgID MessageID
 
 	p.internalSendAsync(ctx, msg, func(ID MessageID, message *ProducerMessage, e error) {
 		err = e
 		msgID = ID
-		wg.Done()
+		wg <- struct{}{}
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go:470]: fill fix it

File: pulsar/producer_partition.go [hunk 6]
```diff
@@ -493,11 +637,100 @@ func (p *partitionProducer) Close() {
...
+	callBackCallAt         time.Time
+}
+
+type onceCallback struct {
+	once     sync.Once
+	callback func(MessageID, *ProducerMessage, error)
+}
+
+func (request *sendRequest) CallBack(messageID MessageID, msg *ProducerMessage, err error) {
+	request.callback.once.Do(func() {
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go:670]: i just want to call the callback once. whatever how many times you call CallBack.

if the context has Done. won't change the origin logic. and just call CallBack as before.

File: pulsar/producer_partition.go [hunk 7]
```diff
@@ -493,11 +637,100 @@ func (p *partitionProducer) Close() {
...
+		request.callback.callback(messageID, msg, err)
+		request.callBackCallAt = time.Now()
+		log.Debugf("gotSemaphoreCost=%v, addToBuilderCost=%v, "+
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go:673]: ok, i'll fix it

File: pulsar/producer_partition.go [hunk 8]
```diff
@@ -343,94 +444,136 @@ func (p *partitionProducer) internalFlush(fr *flushRequest) {
...
 	}
-	p.eventsChan <- sr
 }
 
 func (p *partitionProducer) ReceivedSendReceipt(response *pb.CommandSendReceipt) {
-	pi, ok := p.pendingQueue.Peek().(*pendingItem)
 
-	if !ok {
+	// inject point for mock slow response
+	p.beforeReceiveResponse(response)
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go:534]: this is for mock bad network in unit test. maybe i can remove it from the code.

File: pulsar/producer_partition.go [hunk 9]
```diff
@@ -299,12 +333,57 @@ func (p *partitionProducer) internalSend(request *sendRequest) {
...
+	completed      bool
+	requestCtxDone *bitset.BitSet
+}
+
+// need Lock() before call this func
+func (pi *pendingItem) setFlushAt(flushAt time.Time) {
+	for _, i := range pi.sendRequests {
+		sr := i.(*sendRequest)
+		sr.Lock()
+		sr.flushAt = flushAt
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go:360]: this is just for statistic log in sendRequest Callback. need this var to print debug log.
maybe i can remove it.

File: pulsar/producer_partition.go [hunk 10]
```diff
@@ -299,12 +333,57 @@ func (p *partitionProducer) internalSend(request *sendRequest) {
...
 	sync.Mutex
-	batchData    []byte
-	sequenceID   uint64
-	sendRequests []interface{}
-	completed    bool
+	batchData      []byte
+	sequenceID     uint64
+	sendRequests   []interface{}
+	completed      bool
+	requestCtxDone *bitset.BitSet
```
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go:352]: this is for check if all the sendRequests in pendingItem has been done.
maybe partial or all of it. a bool can't judge partial done case.

File: pulsar/internal/blocking_queue.go [hunk 11]
```diff
@@ -131,6 +139,21 @@ func (bq *blockingQueue) Poll() interface{} {
 	return bq.dequeue()
 }
 
+func (bq *blockingQueue) PollIfSatisfy(condition func(interface{}) bool) (item interface{}, empty bool, satisfy bool) {
```
Comment (merlimat) [inline pulsar/internal/blocking_queue.go:142]: I don't think we need to take messages out individually from the queue. That will easily break ordering. 

Also, we don't need to proactively remove any request from the queue. Just check if the context is expired ***before*** queuing and also when trying to write on the socket. That should be all that we need.

File: pulsar/internal/blocking_queue.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/internal/blocking_queue.go]: removed

File: pulsar/producer_partition.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go]: rename to chanWaitGroup

File: pulsar/producer_partition.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go]: remove defaults here

File: pulsar/producer_partition.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go]: remove the large debug log

File: pulsar/producer_partition.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go]: removed

File: pulsar/producer_partition.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go]: remove the dependency from project,instead we have an bool here

File: pulsar/producer_partition.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (WJL3333) [reply pulsar/producer_partition.go]: remove the whole func, because no debug log now

File: pulsar/internal/blocking_queue.go (hunk 11 above)
Comment (WJL3333) [reply pulsar/internal/blocking_queue.go:142]: yeah, but i seen some use case that we just check with peek and poll after that. this may be not thread safe. if we check and another goroutine poll at the same time. 

and the implementation now is check the context before queuing and also before to write the socket. 

i used to just check the context at those points, but the Send and the SendAsync callback cost much longer time then expected. so there is an external goroutine to check to if the context is Done. and i add some other points  check the context. and we got the expect callback return time near sendtimeout

Comment (WJL3333) [reply pulsar/internal/blocking_queue.go:142]: and there is a new queue in partitionProducer. i use it to check if the sendRequest context is Done. the origin logic in pendingQueue is not changed.

Comment (yarthur1) [reply pulsar/internal/blocking_queue.go:142]: Based on your comit,I have modified something,can you check it?
https://github.com/WJL3333/pulsar-client-go/pull/1

> and there is a new queue in partitionProducer. i use it to check if the sendRequest context is Done. the origin logic in pendingQueue is not changed.

**Comment by yarthur1** [comment]
@WJL3333 
Based on your comit,I have modified something,can you check it?
https://github.com/WJL3333/pulsar-client-go/pull/1

**Comment by WJL3333** [comment]
@merlimat @cckellogg  @jerrypeng  @wolfstudy i think the code works fine now. could you take a look.some logic rewrited.

File: pulsar/producer_partition.go [hunk 12]
```diff
@@ -490,49 +475,77 @@ func (p *partitionProducer) SendAsync(ctx context.Context, msg *ProducerMessage,
...
+	// may be message will block on eventChan publish event
+	checkContextIfBlockOnEventChan := func() (ctxDone bool) {
+		for {
+			select {
+			case p.eventsChan <- sr:
+				return false
+			case <-ctx.Done():
+				log.Debugf("send timeout because for eventChan Block %v", msg)
+				sr.CallBack(nil, msg, ctx.Err())
+				return true
```
Comment (yarthur1) [inline pulsar/producer_partition.go:492]: if go in this ,i think should release publishSemaphore after callback

File: pulsar/producer_partition.go [hunk 13]
```diff
@@ -490,49 +475,77 @@ func (p *partitionProducer) SendAsync(ctx context.Context, msg *ProducerMessage,
...
```
Comment (yarthur1) [inline pulsar/producer_partition.go:528]: the same case with SendAsync

File: pulsar/producer_partition.go [hunk 14]
```diff
@@ -490,49 +475,77 @@ func (p *partitionProducer) SendAsync(ctx context.Context, msg *ProducerMessage,
 	}
```
Comment (yarthur1) [inline pulsar/producer_partition.go:475]: if ctx == nil,  return after callback

File: pulsar/producer_partition.go [hunk 15]
```diff
@@ -322,9 +331,8 @@ func (p *partitionProducer) internalSend(request *sendRequest) {
 				WithField("sequenceID", sequenceID).
 				WithField("properties", msg.Properties).
 				Error("unable to send single message")
```
Comment (yarthur1) [inline pulsar/producer_partition.go]: if fail to add batch builder,should n't relaese publishSemaphore and callback? and why Send individually not retry if add fail?

**Comment by wolfstudy** [comment]
@WJL3333 Can you fix conflict? 

ping @cckellogg @merlimat PTAL again thanks.

**Comment by WJL3333** [comment]
ping @cckellogg @merlimat PTAL. i can reply as soon as posible

**Comment by wolfstudy** [comment]
Move the milestones to 0.2.0

**Comment by wolfstudy** [comment]
@merlimat Can you double-check this change? It seems that this change introduces a new context object, which may conflict with the context we used before.

**Comment by wolfstudy** [comment]
Will close this issue, the #394 has achieved this feature.
